### PR TITLE
accept a description of a connection in the cli and translate it to co…

### PR DESCRIPTION
**Pull request by olasaadi99**
…mmon.ConnectionSet

### CODE REVIEW

File: cmd/analyzer/main.go [hunk 1]
```diff
@@ -83,6 +84,16 @@ func vpcConfigsFromFile(fileName string, inArgs *InArgs) (map[string]*vpcmodel.V
 	return vpcConfigs, nil
 }
 
+func translateCDtoConnectionSet(inArgs *InArgs) *common.ConnectionSet {
+	connection := common.NewConnectionSet(false)
+	if inArgs.Protocol == "icmp" {
+		connection.AddICMPConnection(inArgs.SrcMinPort, inArgs.SrcMaxPort, common.MinPort, common.MaxPort)
```
Comment (ShiriMoran) [inline cmd/analyzer/main.go:90]: Please have the `portMin` and `portMax` for `src` and `dst` be potential part of the input; if it is left unspecified than use the `min` and `max` values as you do now.

File: cmd/analyzer/parse_args.go [hunk 2]
```diff
@@ -18,6 +21,10 @@ type InArgs struct {
 	VPC                   *string
 	Debug                 *bool
 	Version               *bool
+	ConnectionDescription *string
+	Protocol              string
+	SrcMinPort            int64
+	SrcMaxPort            int64
 }
```
Comment (ShiriMoran) [inline cmd/analyzer/parse_args.go:28]: Please add min and max port also for dst
Also, consider adding Q before these new properties or something else so that it will be clear these are relevant only in query mode

File: cmd/analyzer/parse_args.go [hunk 3]
```diff
@@ -18,6 +21,10 @@ type InArgs struct {
 	VPC                   *string
 	Debug                 *bool
 	Version               *bool
+	ConnectionDescription *string
```
Comment (ShiriMoran) [inline cmd/analyzer/parse_args.go:24]: Why is the `ConnectionDescription` required?

File: cmd/analyzer/parse_args.go [hunk 4]
```diff
@@ -160,6 +169,8 @@ func ParseInArgs(cmdlineArgs []string) (*InArgs, error) {
 	args.VPC = flagset.String(VPC, "", "CRN of the VPC to analyze")
 	args.Debug = flagset.Bool(Debug, false, "Run in debug mode")
 	args.Version = flagset.Bool(Version, false, "Prints the release version number")
+	args.ConnectionDescription = flagset.String(ConnectionDescription, "",
+		"Description to a connection in the format: 'Protocol', 'MinSrc', 'MaxSrc'")
```
Comment (ShiriMoran) [inline cmd/analyzer/parse_args.go:173]: Again, why is this required?

File: cmd/analyzer/parse_args.go [hunk 5]
```diff
@@ -180,9 +191,48 @@ func ParseInArgs(cmdlineArgs []string) (*InArgs, error) {
...
+	err = parseConnectionDescription(args.ConnectionDescription, &args)
+	if err != nil {
+		return nil, err
+	}
 
 	return &args, nil
 }
+
+func parseConnectionDescription(connectionDescription *string, args *InArgs) error {
+	if *connectionDescription == "" {
```
Comment (ShiriMoran) [inline cmd/analyzer/parse_args.go:203]: This is not needed. We will soon add `AnalysisType` for query, and then allow specifying these parms only when the mode will be query.
In the meantime just have validity checks that if protocol is specified is legal and that if src/dst min/max ports are specified they are withing range, and that if both min and max are specified then the min is lower than or equal to max

**Review by ShiriMoran — CHANGES_REQUESTED** [review]

File: cmd/analyzer/main.go [hunk 6]
```diff
@@ -83,6 +84,16 @@ func vpcConfigsFromFile(fileName string, inArgs *InArgs) (map[string]*vpcmodel.V
 	return vpcConfigs, nil
 }
 
+func translateCDtoConnectionSet(inArgs *InArgs) *common.ConnectionSet {
+	connection := common.NewConnectionSet(false)
+	if inArgs.Protocol == "icmp" {
```
[the remark this answers is not in the document]
Comment (olasaadi99) [reply cmd/analyzer/main.go:89]: it is supported, I convert it to lower case in parse args..

File: cmd/analyzer/parse_args.go (hunk 4 above)
Comment (olasaadi99) [reply cmd/analyzer/parse_args.go]: though it will be easier for the user to pass single input for the connection description instead of each param (and I can parse it)

Comment (ShiriMoran) [reply cmd/analyzer/parse_args.go]: I see. Perhaps I mislead you. Its better, I think, to use the existing mechanism for feeding parms. This why we can use the existing parsing and validity mechanism. E.g,use the go args mechanism to enforce that ports parm are ints withing range

File: cmd/analyzer/main.go [hunk 7]
```diff
@@ -83,6 +84,18 @@ func vpcConfigsFromFile(fileName string, inArgs *InArgs) (map[string]*vpcmodel.V
 	return vpcConfigs, nil
 }
 
+func translateCDtoConnectionSet(inArgs *InArgs) *common.ConnectionSet {
+	connection := common.NewConnectionSet(false)
+	if common.ProtocolStr(*inArgs.QProtocol) == common.ProtocolICMP {
```
Comment (ShiriMoran) [inline cmd/analyzer/main.go:89]: Please convert to upper case when comparing with  e.g. common.ProtocolICMP

File: cmd/analyzer/main.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (ShiriMoran) [reply cmd/analyzer/main.go]: But now in the fixed version you convert it to upper case?

**Comment by ShiriMoran** [comment]
Much better :-)
To make the parms complete, please add to `AnalysisType` another possible value `query` and add the following validity checks:
1. protocol must be specified for  `AnalysisType` query
2. All the new parms can be specified only if `AnalysisType` is query

File: cmd/analyzer/main.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (olasaadi99) [reply cmd/analyzer/main.go]: right

File: cmd/analyzer/parse_args.go [hunk 8]
```diff
@@ -180,9 +205,74 @@ func ParseInArgs(cmdlineArgs []string) (*InArgs, error) {
...
 	}
+	err = invalidArgsQueryMode(&args, flagset)
+	if err != nil {
+		return nil, err
+	}
 
 	return &args, nil
 }
+
+func isFlagPassed(name string, flagset *flag.FlagSet) bool {
```
Comment (ShiriMoran) [inline cmd/analyzer/parse_args.go:216]: Perhaps remain to `wasFlagSpecified`

File: cmd/analyzer/parse_args.go [hunk 9]
```diff
@@ -180,9 +205,74 @@ func ParseInArgs(cmdlineArgs []string) (*InArgs, error) {
...
+	found := false
+	flagset.Visit(func(f *flag.Flag) {
+		if f.Name == name {
+			found = true
+		}
+	})
+	return found
+}
+
+func isQueryModeParamsPassed(flagset *flag.FlagSet) bool {
```
Comment (ShiriMoran) [inline cmd/analyzer/parse_args.go:226]: Also here, consider renaming to `wereQueryParmSpecified`

File: cmd/analyzer/parse_args.go [hunk 10]
```diff
@@ -180,9 +205,74 @@ func ParseInArgs(cmdlineArgs []string) (*InArgs, error) {
...
+
+	if *args.QDstMinPort > *args.QDstMaxPort {
+		return fmt.Errorf("DstMaxPort %d should be higher than DstMinPort %d", *args.QSrcMaxPort, *args.QSrcMinPort)
+	}
+
+	if *args.QDstMinPort > common.MaxPort || *args.QDstMinPort < common.MinPort ||
+		*args.QDstMaxPort > common.MaxPort || *args.QDstMaxPort < common.MinPort {
+		return fmt.Errorf("DstMaxPort and DstMinPort must be in ranges [%d, %d]", common.MinPort, common.MaxPort)
+	}
+
```
Comment (ShiriMoran) [inline cmd/analyzer/parse_args.go:253]: Please consider having the checks for `dst/src` in a function called once, to avoid code duplication

**Comment by ShiriMoran** [comment]
Looks good, a few more minor comments

**Review by ShiriMoran — CHANGES_REQUESTED** [review]

File: pkg/vpcmodel/output.go [hunk 11]
```diff
@@ -38,6 +38,7 @@ const (
 	AllSubnetsNoPGW                      // connectivity between subnets (consider nacl only)
 	SubnetsDiff                          // diff between subnets connectivity of two cfgs (consider nacl + pgw)
 	EndpointsDiff                        // diff between vsis connectivity of two cfgs
+	Query                                // analyze a connection description
```
Comment (adisos) [inline pkg/vpcmodel/output.go:41]:
```suggestion
	Explain                                // explain specified connectivity, given src,dst and connection 
```

File: cmd/analyzer/parse_args.go [hunk 12]
```diff
@@ -59,6 +76,7 @@ const (
 	singleSubnet     = "single_subnet"      // single subnet connectivity analysis
 	allEndpointsDiff = "diff_all_endpoints" // semantic diff of allEndpoints analysis between two configurations
 	allSubnetsDiff   = "diff_all_subnets"   // semantic diff of allSubnets analysis between two configurations
+	queryMode        = "query"              // analyze a connection description
```
Comment (adisos) [inline cmd/analyzer/parse_args.go:79]:
```suggestion
	explainMode        = "explain"              // explain specified connectivity
```

**Review by adisos — CHANGES_REQUESTED** [review]
Please rename "query" analysis type to be "explain" instead (@ShiriMoran  confirmed the renaming does make sense).
Added few examples in comments where to apply this change, go over all relevant places to change.

File: cmd/analyzer/main.go [hunk 13]
```diff
@@ -83,6 +86,18 @@ func vpcConfigsFromFile(fileName string, inArgs *InArgs) (map[string]*vpcmodel.V
...
+		connection.AddICMPConnection(*inArgs.QSrcMinPort, *inArgs.QSrcMaxPort, *inArgs.QDstMinPort, *inArgs.QDstMaxPort)
```
Comment (adisos) [inline cmd/analyzer/main.go:92]: why are src/dst ports sent to `AddICMPConnection` as icmp type and code? 
initially for simplicity, we do not have input icmp type / code, so should add ALL ICMP connections (use the defualt min and max type/code values from the `common` pkg).

File: cmd/analyzer/parse_args.go [hunk 14]
```diff
@@ -180,9 +205,78 @@ func ParseInArgs(cmdlineArgs []string) (*InArgs, error) {
...
+func PortInRange(port int64) bool {
+	if port > common.MaxPort || port < common.MinPort {
+		return false
+	}
+
+	return true
+}
+
+func validRangeConnectionExplainMode(args *InArgs) error {
+	if *args.QSrcMinPort > *args.QSrcMaxPort {
```
Comment (ShiriMoran) [inline cmd/analyzer/parse_args.go:244]: minPort and maxPort may be equal

File: cmd/analyzer/parse_args.go [hunk 15]
```diff
@@ -180,9 +205,78 @@ func ParseInArgs(cmdlineArgs []string) (*InArgs, error) {
...
+		return fmt.Errorf("srcMaxPort %d should be higher than srcMinPort %d", *args.QSrcMaxPort, *args.QSrcMinPort)
+	}
+
+	if *args.QDstMinPort > *args.QDstMaxPort {
```
Comment (ShiriMoran) [inline cmd/analyzer/parse_args.go:248]: Also here, minPort and maxPort may be equal

**Review by ShiriMoran — CHANGES_REQUESTED** [review]
Almost done.
Just one comment: please let me mark as resolve my comments, so that I can easily follow them

**Review by ShiriMoran — APPROVED** [review]
